src/dev_usb_freebsd.c

```diff
@@ -0,0 +1,324 @@
+				if (rdbytes > 6)
+					curr_pos[2] = iev[5] | (iev[6] << 8);
+				if (rdbytes > 8)
+					curr_pos[3] = iev[5] | (iev[6] << 8);
+				if (rdbytes > 10)
+					curr_pos[4] = iev[5] | (iev[6] << 8);
+				if (rdbytes > 12)
+					curr_pos[5] = iev[5] | (iev[6] << 8);
+				return axis_event(inp, last_pos, curr_pos);
+			case 2: // Three axis... rX, rY, rZ
+				flush = true;
+				curr_pos[3] = iev[1] | (iev[2] << 8);
+				curr_pos[4] = iev[3] | (iev[4] << 8);
+				curr_pos[5] = iev[5] | (iev[6] << 8);
+				return axis_event(inp, last_pos, curr_pos);
+			case 3: // Button change event.
+				flush = true;
+				curr_buttons = iev[1] | (iev[2] << 8) | (iev[3] << 16);
+				if (last_buttons != curr_buttons) {
+					last_buttons = button_event(inp, last_buttons, curr_buttons);
+					return 0;
+				}
+				return -1;
+			case 23: // Battery char level
+				logmsg(LOG_INFO, "Battery level: %%%d\n", iev[1]);
+				break;
+			default:
+				if(verbose) {
+					logmsg(LOG_DEBUG, "unhandled event: %d\n", iev[0]);
+				}
+				break;
+		}
+	}
+
+	return -1;
+}
+
+int open_dev_usb(struct device *dev)
+{
+	if((dev->fd = open(dev->path, O_RDWR)) == -1) {
+		if((dev->fd = open(dev->path, O_RDONLY)) == -1) {
+			logmsg(LOG_ERR, "failed to open device: %s\n", strerror(errno));
+			return -1;
+		}
+		logmsg(LOG_WARNING, "opened device read-only, LEDs won't work\n");
+	}
+	dev->num_axes = AXES;
+#if NEED_STUFF
+	/*
+	 * TODO: Fetch these...
+	 * - num_axes needs to be a bit complex, there's two reports each with three axis...
+	 * - num_buttons is trivial though.
+	 * - min/max is easy, not sure what fuzz even is.
+	 * - Also, it seems nothing actually uses these things...
+	 */
+	dev->num_buttons = 21;
+	dev->minval = malloc(dev->num_axes * sizeof *dev->minval);
+	dev->maxval = malloc(dev->num_axes * sizeof *dev->maxval);
+	dev->fuzz = malloc(dev->num_axes * sizeof *dev->fuzz);
+
+	if(!dev->minval || !dev->maxval || !dev->fuzz) {
+		free(dev->minval);
+		free(dev->maxval);
+		free(dev->fuzz);
+		logmsg(LOG_ERR, "failed to allocate memory: %s\n", strerror(errno));
+		return -1;
+	}
+
+	/* if the device is an absolute device, find the minimum and maximum axis values */
+	for(int i=0; i<dev->num_axes; i++) {
+		dev->minval[i] = DEF_MINVAL;
+		dev->maxval[i] = DEF_MAXVAL;
+		dev->fuzz[i] = 0;
+	}
+#endif
+
+	if(cfg.led == LED_ON || (cfg.led == LED_AUTO && first_client())) {
+		set_led_hid(dev, 1);
+	} else {
+		/* Some devices start with the LED enabled, make sure to turn it off
+		 * explicitly if necessary.
+		 *
+		 * XXX G.Ebner reports that some devices (SpaceMouse Compact at least)
+		 * fail to turn their LED off at startup if it's not turned explicitly
+		 * on first. We'll need to investigate further, but it doesn't seem to
+		 * cause any visible blinking, so let's leave the redundant call to
+		 * enable it first for now. See github pull request #39:
+		 * https://github.com/FreeSpacenav/spacenavd/pull/39
+		 */
+		set_led_hid(dev, 1);
+		set_led_hid(dev, 0);
+	}
```

Comment:
I don't like the duplication of this logic, but it can be merged like this, and I'll factor it out later.

Reply:
Yeah, I just don't have enough devices to be able to test this reasonably well.